### ISSUE TITLE
Shared key

### ISSUE THREAD

**Issue by unparagoned** [frawau/aiotuya#1]
There should be a way of doing a PM. Anyway really nice script, it would really help and I don't know this but I herd that all devices provisioned by a key can then be controlled by that key. Have a read through the following thread where some people decided against releasing an app with the key already loaded for users to use. 

https://github.com/codetheweb/tuyapi/issues/20

**Comment by frawau** [comment]
Hi,

Well, on the LAN side, each device gets its own key. On the cloud side, I am not entirely sure, but since the provisioning requires you to login into the Tuya Cloud with your email and password, I would think that you need both the key and the login to control a device.

In any case, this key and id is associated with my application and I intend to block internet access to the devices once they are provisioned.   So I feel it is safe enough.

**Comment by unparagoned** [comment]
I don't know too much about it but I thought the risk was more to people that use your script. This post by @dlashua was the one I was referring to. I don't know if it applies to your situation.
https://github.com/codetheweb/tuyapi/issues/20#issuecomment-365991745


[dlashua](/dlashua)

> That's perfect! I applied for the Cloud API Key 3 days ago so that I could help get this underway, but I've yet to get info from them and my application still shows as "under review".
> 
> It's a shame they didn't make this a little bit easier. Tuya products are EVERYWHERE, they work quite well (as long as you use their permissions laiden apps), and are fairly inexpensive. And all that's "missing" is some documentation, and a way to easily access localKey.
> 
> There are several big downsides to the cloud approach as Tuya presents it. As best as I can tell, you'll need to perform a "tuya.cloud.user.sync" for each user that wishes to use the cloud through your key. Additionally, at that point, that cloud account will have access (without a password, from that point on) to that user's devices. So, this means 1) you can't put the Cloud API creds directly in the library or else everyone will have access to everyone else's devices, 2) because of this, a intermediate API will need to be developed for the library to hit, 3) this API will have to be hosted somewhere ($$$) and publically available, 4) users will have to trust this cloud service with their credentials.
> 
> Another option will be to require that every user seek their own Cloud API credentials, but, as you can see from my experience, this doesn't cater to the "I want it now" mentality, as I've been waiting three days with no response.

**Comment by frawau** [comment]
Admittedly, I did not focus my attention to the cloud side of things, my aim being local/LAN control of the devices.  

Still my understanding was different. The cloud access needs 2 bits of information, the application key and sercret and a login name and password. Each user should have her own login name and password. This control which devices you have access to. The API key and secret are there to ensure that the application you use is legit.  

So the application key and secret will allow to issue legit requests to the cloud, but you still need the email and password to access the devices.

That is my understanding anyway.... 

What I saw was reluctance to share keys and secret that had been lifted from a phone application.
But this does not apply to my case, I applied for an app and got the keys and secret.

Regards,
    François

**Comment by jharrang** [comment]
Hey François,

This looks really promising! I'm going to try to wire this up to work with homeassistant since I used to have a custom component there that worked with a sniffed local_key and device_id, but switched to the current builtin component for ease of adding devices. The old local-control way was **significantly** faster at certain times of the day, so it will be great to get back to that.

At some point I actually applied for API access and got approved, but I just wanted to verify that I can likely use my AccessId and AccessKey as a drop-in-replacement for yours in provision.py. You just registered for the "Basic Version" cloud service at https://developer.tuya.com/user/cloud and then used the Cloud API info from that, correct?

Best,
James

**Comment by frawau** [comment]
Hi James,

Yes I applied and got some "Basic Version" API key by creating an application and using that application key (The Android one I think).  If you got one like that, you probably can just use yours.  

One think I did notice though, was that at some point I tried to use their Chinese server but it did not work.... It was early on so it might have been a mistake on my part, but I did not really investigate and used the American server after that. Just be aware of that if you use Chinese or European servers.

Cheers,
    François
